go/vt/grpcclient/zstd.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpcclient
+
+import (
+	"io"
+
+	"github.com/klauspost/compress/zstd"
+)
+
+// ZstdCompressor is a gRPC compressor using zstd
+type ZstdCompressor struct{}
+
+// Name is "zstd"
+func (s ZstdCompressor) Name() string {
+	return "zstd"
+}
+
+// Compress wraps with a ZstdReader
+func (s ZstdCompressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	enc, err := zstd.NewWriter(w)
+	if err != nil {
+		return nil, err
+	}
+	return enc, nil
+}
+
+// Decompress wraps with a ZstdReader
+func (s ZstdCompressor) Decompress(r io.Reader) (io.Reader, error) {
+	dec, err := zstd.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+	return dec, nil
+}
```

Comment:
This will be relatively slow as it uses the "streaming" method which is targeted at processing large chunks of data vs smaller in-memory buffers. You can see the two methods used here: https://github.com/vitessio/vitess/pull/12950/files#diff-13b5d05ff0ac248bd9159f5c69d4e1b02db82aab71b2db234030afe033d2f5e1

And discussed here: https://github.com/klauspost/compress/tree/master/zstd#decompressor

We may also want to increase the default concurrency based on GOMAXPROCS. You can see that discussed here: https://github.com/klauspost/compress/tree/master/zstd#usage

I bring this up as gRPC is fairly performance sensitive and I think that's why we're adding support for ZSTD here, right?

Reply:
Good call out Matt. I wasn't aware of the two modes offered by the library. To provide more context, we got a nice win by switching to Zstd compression on memcached, so we admittedly hastily wrote up this patch to experiment with Zstd on Vitess. We confirmed the patch works in dev but did no meaningful perf testing.

Going by the docs I agree non-streaming/buffer mode sounds like it makes more sense. We can take a shot at switching over to that.

Re: concurrency I am a bit fuzzy on this line in the docs:

> The buffer decoder does everything on the same goroutine and does nothing concurrently. It can however decode several buffers concurrently. Use WithDecoderConcurrency(n) to limit that.

I'm guessing that means "does nothing in parallel (1 goroutine) but can decode concurrently (in the coroutine-sense)".

I'll encourage whoever picks this up next on our side to actually look at `klauspost/compress` internals as well as the calling context of this code in Vitess to be a little more deliberate about concurrency, defaults, and what tunables we may want to expose.